File: ts/src/advisor.ts
import type { Advice } from "./types.js";

export interface Advisor {
  generateAdvice(context: Record<string, unknown>): Promise<Advice>;
  modelId(): string;
}

export class RemoteAdvisor implements Advisor {
  constructor(
    private endpoint: string,
    private timeoutMs = 5000,
  ) {}

  async generateAdvice(context: Record<string, unknown>): Promise<Advice> {
    const resp = await fetch(`${this.endpoint}/advise`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(context),
      signal: AbortSignal.timeout(this.timeoutMs),
    });
    if (!resp.ok) throw new Error(`advisor error: ${resp.status}`);
    const data = await resp.json();
    return {
      domain: data.domain,
      steering_text: data.steering_text,
      confidence: data.confidence,
      constraints: data.constraints ?? [],
      metadata: data.metadata ?? {},
      version: data.version ?? "1.0",
    };
  }

  modelId(): string {
    return `remote:${this.endpoint}`;
  }

  static fromEndpoint(url: string): RemoteAdvisor {
    return new RemoteAdvisor(url);
  }
}

File: py/src/fit/advisor.py
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any

from .types import Advice


class Advisor(ABC):
    """Loads a trained advisor and generates per-instance advice."""

    @abstractmethod
    def generate_advice(self, context: dict[str, Any]) -> Advice:
        """Generate steering advice for the given context."""

    @abstractmethod
    def model_id(self) -> str:
        """Return advisor model identifier."""


class RemoteAdvisor(Advisor):
    """Advisor backed by a remote inference endpoint."""

    def __init__(self, endpoint: str, timeout_ms: int = 5000) -> None:
        self._endpoint = endpoint
        self._timeout_ms = timeout_ms

    def generate_advice(self, context: dict[str, Any]) -> Advice:
        import httpx
        import json

        resp = httpx.post(
            f"{self._endpoint}/advise",
            json=context,
            timeout=self._timeout_ms / 1000,
        )
        resp.raise_for_status()
        data = resp.json()
        return Advice(
            domain=data["domain"],
            steering_text=data["steering_text"],
            confidence=data["confidence"],
            constraints=data.get("constraints", []),
            metadata=data.get("metadata", {}),
        )

    def model_id(self) -> str:
        return f"remote:{self._endpoint}"

    @classmethod
    def from_endpoint(cls, url: str) -> RemoteAdvisor:
        return cls(endpoint=url)
